File: app/src/main/java/com/riders/thelab/core/utils/LabTypefaceManager.java
package com.riders.thelab.core.utils;

import android.content.Context;
import android.graphics.Typeface;

import java.lang.reflect.Field;
import java.util.HashMap;
import java.util.Map;

import timber.log.Timber;

public class LabTypefaceManager {

    private LabTypefaceManager() {
    }

    /**
     * Using reflection to override default typeface
     * NOTICE: DO NOT FORGET TO SET TYPEFACE FOR APP THEME AS DEFAULT TYPEFACE WHICH WILL BE OVERRIDDEN
     *
     * @param context                    to work with assets
     * @param defaultFontNameToOverride  for example "monospace"
     * @param customFontFileNameInAssets file name of the font from assets
     */
    public static void overrideFont(Context context,
                                    String defaultFontNameToOverride,
                                    String customFontFileNameInAssets) {

        final Typeface customFontTypeface =
                Typeface.createFromAsset(context.getAssets(), customFontFileNameInAssets);

        if (LabCompatibilityManager.isLollipop()) {
            Map<String, Typeface> newMap = new HashMap<String, Typeface>();
            newMap.put("serif", customFontTypeface);
            try {
                final Field staticField = Typeface.class
                        .getDeclaredField("sSystemFontMap");
                staticField.setAccessible(true);
                staticField.set(null, newMap);
            } catch (NoSuchFieldException e) {
                e.printStackTrace();
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            }
        } else {
            try {
                final Field defaultFontTypefaceField =
                        Typeface.class.getDeclaredField(defaultFontNameToOverride);
                defaultFontTypefaceField.setAccessible(true);
                defaultFontTypefaceField.set(null, customFontTypeface);
            } catch (Exception e) {
                Timber.e(
                        "Can not set custom font " + customFontFileNameInAssets
                                + " instead of " + defaultFontNameToOverride);
            }
        }
    }

}

File: app/src/main/java/com/riders/thelab/core/utils/LabTypefaceManager.kt
package com.riders.thelab.core.utils

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Typeface
import timber.log.Timber
import java.util.*

class LabTypefaceManager private constructor() {

    companion object {

        /**
         * Using reflection to override default typeface
         * NOTICE: DO NOT FORGET TO SET TYPEFACE FOR APP THEME AS DEFAULT TYPEFACE WHICH WILL BE OVERRIDDEN
         *
         * @param context                    to work with assets
         * @param defaultFontNameToOverride  for example "monospace"
         * @param customFontFileNameInAssets file name of the font from assets
         */
        @SuppressLint("BinaryOperationInTimber")
        fun overrideFont(context: Context,
                         defaultFontNameToOverride: String,
                         customFontFileNameInAssets: String) {

            val customFontTypeface = Typeface.createFromAsset(context.assets, customFontFileNameInAssets)

            if (LabCompatibilityManager.isLollipop()) {
                val newMap: MutableMap<String, Typeface> = HashMap()
                newMap["serif"] = customFontTypeface
                try {
                    val staticField = Typeface::class.java
                            .getDeclaredField("sSystemFontMap")
                    staticField.isAccessible = true
                    staticField[null] = newMap
                } catch (e: NoSuchFieldException) {
                    e.printStackTrace()
                } catch (e: IllegalAccessException) {
                    e.printStackTrace()
                }
            } else {
                try {
                    val defaultFontTypefaceField = Typeface::class.java.getDeclaredField(defaultFontNameToOverride)
                    defaultFontTypefaceField.isAccessible = true
                    defaultFontTypefaceField[null] = customFontTypeface
                } catch (e: Exception) {
                    Timber.e(
                            "Can not set custom font " + customFontFileNameInAssets
                                    + " instead of " + defaultFontNameToOverride)
                }
            }
        }

    }
}